contracts/adapters/IPRebalanceExtension.sol

```diff
@@ -0,0 +1,478 @@
+            _transformComponent,
+            unitsToUntransform,
+            _untransformData
+        );
+
+        invokeManager(module, callData);
+    }
+
+    /**
+     * Untransforms a component
+     */
+    function _transform(address _transformComponent, bytes memory _transformData, bool _transformRemaining) internal {
+        require(tradesComplete, "trades not complete");
+
+        TransformInfo memory transformInfo = transformComponentInfo[_transformComponent];
+
+        require(transformInfo.minTransformDelay + lastTransform[_transformComponent] < block.timestamp, "delay not elapsed");
+        lastTransform[_transformComponent] = block.timestamp;
+
+        require(transformInfo.underlyingComponent != address(0), "nothing to transform");
+        require(
+            transformInfo.transformHelper.shouldTransform(transformInfo.underlyingComponent, _transformComponent),
+            "transform unavailable"
+        );
+
+        uint256 unitsToTransform;
+        uint256 currentRawUnderlying = setToken.getDefaultPositionRealUnit(transformInfo.underlyingComponent).toUint256();
+        uint256 maxTransformUnits = transformInfo.maxTransformSize.preciseDiv(setToken.totalSupply());
+
+        if (_transformRemaining) {
+            require(maxTransformUnits >= currentRawUnderlying, "transform units greater than max");
+            unitsToTransform = currentRawUnderlying;
+        } else {
+            uint256 currentUnits = setToken.getDefaultPositionRealUnit(_transformComponent).toUint256();
+            uint256 exchangeRate = transformInfo.transformHelper.getExchangeRate(transformInfo.underlyingComponent, _transformComponent);
+
+            uint256 currentUnitsUnderlying = currentUnits.preciseDiv(exchangeRate);
+            uint256 targetUnitsUnderlying = rebalanceParams[_transformComponent];
+
+            unitsToTransform = targetUnitsUnderlying > currentUnitsUnderlying ? targetUnitsUnderlying.sub(currentUnitsUnderlying) : 0;
+
+            if (unitsToTransform > currentRawUnderlying) {
+                unitsToTransform = currentRawUnderlying;
+            }
+        }
+
+        if (unitsToTransform > maxTransformUnits) {
+            unitsToTransform = maxTransformUnits;
+        }
+
+
+        require(unitsToTransform > 0, "nothing to transform");
+
+        (address module, bytes memory callData) = transformInfo.transformHelper.getTransformCall(
+            manager.setToken(),
+            transformInfo.underlyingComponent,
+            _transformComponent,
+            unitsToTransform,
+            _transformData
+        );
+
+        invokeManager(module, callData);
+    }
+
+    /**
+     * Parameterizes the rebalancing trades through GeneralIndexModule. For all transform components, the target units
+     * remain fixed at their current allocations. For non-transform units, the target units are calculated by taking into
+     * account the total amount of the component needed to create the target trasnform units.
```

Comment:
What are all the components that make up the target non-transform units? There's some equation here I should be able to check against